cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/TimeScheduler.java

```diff
@@ -291,6 +293,32 @@ private void scheduleJob(TriggerKey triggerKey, String scheduleName, String cron
     return getScheduledRuntime(program, programType, false);
   }
 
+  public List<ScheduledRuntime> nextScheduledRuntime(ProgramId program, SchedulableProgramType programType,
+                                                     long startTimeSecs, long endTimeSecs)
+    throws SchedulerException {
+    // decrease the start time by one second to include the next fire time that is equal to the start time
+    Date startTime = new Date(TimeUnit.SECONDS.toMillis(startTimeSecs - 1));
+    long endTimeMillis = TimeUnit.SECONDS.toMillis(endTimeSecs);
+    List<ScheduledRuntime> scheduledRuntimes = new ArrayList<>();
+    try {
+      for (Trigger trigger : scheduler.getTriggersOfJob(jobKeyFor(program, programType))) {
+        if (scheduler.getTriggerState(trigger.getKey()) == Trigger.TriggerState.PAUSED) {
+          // if the trigger is paused, then skip getting the next fire time
+          continue;
+        }
+        Date nextFireTime = trigger.getFireTimeAfter(startTime);
+        while (nextFireTime != null && nextFireTime.getTime() < endTimeMillis) {
+          ScheduledRuntime runtime = new ScheduledRuntime(trigger.getKey().toString(), nextFireTime.getTime());
+          scheduledRuntimes.add(runtime);
+          nextFireTime = trigger.getFireTimeAfter(nextFireTime);
+        }
+      }
+    } catch (org.quartz.SchedulerException e) {
+      throw new SchedulerException(e);
```

Comment:
why do you have to wrap

Reply:
it's not really necessary here, but it is following the style of other methods. When methods in `TimeScheduler` is called in `CoreSchedulerService`, we want to handle scheduler related exceptions separately, so we wrap scheduler related exceptions in `SchedulerException`